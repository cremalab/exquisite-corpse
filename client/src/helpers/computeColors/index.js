// @flow

import Color from 'tinycolor2'

type ColorMap = {
  [color: string | number]: string
}

type ConfigObj = {
  steps: number,
  shift: number
}

export default (colors: ColorMap, { steps = 10, shift = 40 }: ConfigObj = {}): Object =>
  Object.keys(colors).reduce((acc, color) => {
    const colorValue = colors[color]
    const computedColors = [...Array(steps).keys()].reduce(
      (acc, step) => {
        const currentStep  = step + 1
        const currentColor = Color(colorValue).toString()
        const amount       = currentStep * (shift / steps)
        const tint         = Color(colorValue).lighten(amount).toString()
        const shade        = Color(colorValue).darken(amount).toString()
        const white        = tint === '#ffffff'
        const black        = shade === '#000000'
        return {
          ...acc,
          [color]: currentColor,
          ...(white ? {} : { [`${color}-tint-${currentStep}`]: tint }),
          ...(black ? {} : { [`${color}-shade-${currentStep}`]: shade }),
        }
      },
      {}
    )
    return ({...acc, ...computedColors})
  }, {})
