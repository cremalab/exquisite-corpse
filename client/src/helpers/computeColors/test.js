// @flow
import computeColors from '.'

const colors = {
  primary:   '#79589f',
  secondary: '#596981',
  tertiary:  '#e3e7ef',
  blue:      '#408fec',
  white:     '#FFFFFF',
}

describe('computeColors', () => {
  it('works', () => {
    const actual = computeColors(colors)
    expect(actual).toMatchSnapshot()
  })
})
