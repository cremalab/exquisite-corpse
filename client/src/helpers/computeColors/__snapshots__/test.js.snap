// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeColors works 1`] = `
Object {
  "blue": "#408fec",
  "blue-shade-1": "#2d84ea",
  "blue-shade-10": "#092d57",
  "blue-shade-2": "#1b79e8",
  "blue-shade-3": "#166fd9",
  "blue-shade-4": "#1466c7",
  "blue-shade-5": "#125cb4",
  "blue-shade-6": "#1053a2",
  "blue-shade-7": "#0e498f",
  "blue-shade-8": "#0c407c",
  "blue-shade-9": "#0b366a",
  "blue-tint-1": "#539aee",
  "blue-tint-10": "#fafcfe",
  "blue-tint-2": "#65a5f0",
  "blue-tint-3": "#78b0f2",
  "blue-tint-4": "#8abbf3",
  "blue-tint-5": "#9dc5f5",
  "blue-tint-6": "#afd0f7",
  "blue-tint-7": "#c2dbf9",
  "blue-tint-8": "#d4e6fb",
  "blue-tint-9": "#e7f1fd",
  "primary": "#79589f",
  "primary-shade-1": "#6f5192",
  "primary-shade-10": "#150f1c",
  "primary-shade-2": "#654985",
  "primary-shade-3": "#5b4278",
  "primary-shade-4": "#513b6a",
  "primary-shade-5": "#47345d",
  "primary-shade-6": "#3d2c50",
  "primary-shade-7": "#332543",
  "primary-shade-8": "#291e36",
  "primary-shade-9": "#1f1729",
  "primary-tint-1": "#8363a9",
  "primary-tint-10": "#e1d9ea",
  "primary-tint-2": "#8e70b0",
  "primary-tint-3": "#987db7",
  "primary-tint-4": "#a28abe",
  "primary-tint-5": "#ad97c6",
  "primary-tint-6": "#b7a4cd",
  "primary-tint-7": "#c2b2d4",
  "primary-tint-8": "#ccbfdb",
  "primary-tint-9": "#d6cce3",
  "secondary": "#596981",
  "secondary-shade-1": "#515f75",
  "secondary-shade-10": "#060708",
  "secondary-shade-2": "#485569",
  "secondary-shade-3": "#404c5d",
  "secondary-shade-4": "#384251",
  "secondary-shade-5": "#2f3845",
  "secondary-shade-6": "#272e39",
  "secondary-shade-7": "#1f242c",
  "secondary-shade-8": "#161a20",
  "secondary-shade-9": "#0e1114",
  "secondary-tint-1": "#61738d",
  "secondary-tint-10": "#cbd1db",
  "secondary-tint-2": "#6a7d98",
  "secondary-tint-3": "#7687a1",
  "secondary-tint-4": "#8292a9",
  "secondary-tint-5": "#8f9db1",
  "secondary-tint-6": "#9ba7ba",
  "secondary-tint-7": "#a7b2c2",
  "secondary-tint-8": "#b3bcca",
  "secondary-tint-9": "#bfc7d3",
  "tertiary": "#e3e7ef",
  "tertiary-shade-1": "#d6dce8",
  "tertiary-shade-10": "#6178a5",
  "tertiary-shade-2": "#c9d1e0",
  "tertiary-shade-3": "#bcc6d9",
  "tertiary-shade-4": "#afbad1",
  "tertiary-shade-5": "#a2afca",
  "tertiary-shade-6": "#95a4c2",
  "tertiary-shade-7": "#8899bb",
  "tertiary-shade-8": "#7b8eb4",
  "tertiary-shade-9": "#6e83ac",
  "tertiary-tint-1": "#f0f2f6",
  "tertiary-tint-2": "#fdfdfe",
  "white": "#ffffff",
  "white-shade-1": "#f5f5f5",
  "white-shade-10": "#999999",
  "white-shade-2": "#ebebeb",
  "white-shade-3": "#e0e0e0",
  "white-shade-4": "#d6d6d6",
  "white-shade-5": "#cccccc",
  "white-shade-6": "#c2c2c2",
  "white-shade-7": "#b8b8b8",
  "white-shade-8": "#adadad",
  "white-shade-9": "#a3a3a3",
}
`;
